Silicon/AlderlakePkg/Library/BaseSerialIoI2cLib/SerialIoI2cLib.c

```diff
@@ -0,0 +1,374 @@
+      SerialIoI2cDisableController (MmioBase);
+      SerialIoI2cInitializeWriteRead (MmioBase, SlaveAddress, &Frequency, TimingConfig);
+      return EFI_DEVICE_ERROR;
+    }
+
+    //
+    // Add Data to TX FIFO
+    // If TX FIFO is Full or at its limit decrease Timeout
+    //
+    if (SerialIoI2cPrivateWrite (&WriteLength, &WriteBuffer, MmioBase, Stop, Restart)) {
+      Restart = FALSE;
+    } else {
+      MicroSecondDelay (1);
+      Timeout--;
+    }
+
+    if (WriteLength == 0) {
+      Status = SerialIoI2cWaitTxFifoEmptyIcNotActive (MmioBase);
+      if (Status != EFI_SUCCESS) {
+        DEBUG ((DEBUG_ERROR, "SerialIoI2cWrite Error Tx FIFO not empty or still active\n"));
+        return Status;
+      }
+      return EFI_SUCCESS;
+    }
+
+    if (Timeout == 0) {
+      SerialIoI2cDisableController (MmioBase);
+      SerialIoI2cInitializeWriteRead (MmioBase, SlaveAddress, &Frequency, TimingConfig);
+      DEBUG ((DEBUG_ERROR, "SerialIoI2cWrite Timeout\n"));
+      return EFI_TIMEOUT;
+    }
+  }
+}
+
+/**
+  Execute an I2C read command
+
+  Addressing mode will defined based upon SlaveAddress and mask:
+  SlaveAdress & 0x80000000 - if true will result in 10bit addressing
+  Refer to: I2C_ADDRESSING_10_BIT
+
+  @param[in]   PciCfgBase     Pci Config Base Address of specified I2C controller
+                              @Note: If I2C Number is provided this parameter is ignored and can be left as 0
+  @param[in]   I2cNumber      Optional: I2C controller applicable on bus 0 only
+  @param[in]   SlaveAddress  -Address of the end point device
+  @param[in]   ReadLength     Read length of the current transaction
+  @param[in]   ReadBuffer     Pointer to data in read buffer
+  @param[in]   TimeBudget     Time allotted to verify if I2C is enabled
+  @param[in]   Frequency      Bus clock frequency (in kHz)
+  @param[in]   TimingConfig   Optional: Pointer to struct passed down to override SDA Hold and SCL LCNT/HCNT settings
+  @param[in]   Restart        Indication if restart is needed, if TRUE will Initialize I2C controller.
+                              @Note: Restart is required in first transaction.
+
+  @retval   EFI_INVALID_PARAMETER   Invalid length / buffer parameters were passed into this function
+  @retval   EFI_TIMEOUT             Timeout occurred in I2C disable/enable or wrong ENST value
+  @retval   EFI_DEVICE_ERROR        The verification failed
+  @retval   EFI_SUCCESS             Read was successful
+**/
+EFI_STATUS
+SerialIoI2cRead (
+  IN UINTN                 PciCfgBase,
+  IN UINT8                 *I2cNumber  OPTIONAL,
+  IN UINT32                SlaveAddress,
+  IN UINT32                ReadLength,
+  IN UINT8                 *ReadBuffer,
+  IN UINT64                TimeBudget,
+  IN UINTN                 Frequency,
+  IN SERIAL_IO_I2C_TIMING  *TimingConfig  OPTIONAL,
+  IN BOOLEAN               Restart
+  )
+{
+  EFI_STATUS  Status;
+  UINT64      Timeout;
+  UINTN       MmioBase;
+  //I2C Post Code
+  UINT64 MmioBase1;
```

Comment:
Can't this function be implemented without mioBase1 variable?

Reply:
The reason why I introduced that variable is to avoid the casts.